source/dofs/dof_handler.cc

```diff
@@ -997,7 +997,10 @@ void DoFHandler<dim,spacedim>::distribute_dofs (const FiniteElement<dim,spacedim
   Assert (tria->n_levels() > 0,
           ExcMessage("The Triangulation you are using is empty!"));
 
-  fe_collection = std_cxx14::make_unique<hp::FECollection<dim, spacedim>>(ff);
+  // Only recreate the FECollection if we don't already store an instance
+  // of the FiniteElement class given.
+  if (fe_collection == nullptr || typeid((*fe_collection)[0]) != typeid(ff))
```

Comment:
that's not good enough. You wouldn't be able to distinguish `FE_Q(1)` from `FE_Q(2)` this way.

I think you want to just compare the actual pointers. I.e., if someone calls
```
  dof_handler.distribute_dofs (dof_handler.get_fe());
```
then avoid the cloning. That's in essence what @luca-heltai is trying to do, after all.